strings/kmp/kmp.go

```diff
@@ -1,117 +1,49 @@
-// 	var text string
-// 	var word string
-
-// 	if isTakingInputFromCommandLine { // case of command line input
-// 		args := os.Args
-// 		if len(args) <= 2 {
-// 			log.Fatal("Not enough arguments. Two string arguments separated by spaces are required!")
-// 		}
-// 		word = args[1]
-// 		text = args[2]
-// 		for i := 3; i < len(args); i++ {
-// 			text = text + " " + args[i]
-// 		}
-// 	} else { // case of file input
-// 		patFile, err := ioutil.ReadFile("../pattern.txt")
-// 		if err != nil {
-// 			log.Fatal(err)
-// 		}
-// 		textFile, err := ioutil.ReadFile("../text.txt")
-// 		if err != nil {
-// 			log.Fatal(err)
-// 		}
-// 		text = string(textFile)
-// 		word = string(patFile)
-// 	}
+// Kmp Function kmp performing the Knuth-Morris-Pratt algorithm.
+func Kmp(word, text string, patternTable []int) []int {
+	if len(word) > len(text) {
+		return nil
+	}
 
-// 	if len(word) > len(text) {
-// 		log.Fatal("Pattern is longer than text!")
-// 	}
-// 	fmt.Printf("\nRunning: Knuth-Morris-Pratt algorithm.\n\n")
-// 	fmt.Printf("Search word (%d chars long): %q.\n", len(word), word)
-// 	fmt.Printf("Text        (%d chars long): %q.\n\n", len(text), text)
+	var (
+		i, j    int
+		matches []int
+	)
+	for i+j < len(text) {
 
-// 	r := kmp(text, word)
-// 	if r.resultPosition == notFoundPosition {
-// 		fmt.Printf("\n\nWord was not found.\n%d comparisons were done.", r.numberOfComparison)
-// 	} else {
-// 		fmt.Printf("\n\nWord %q was found at position %d in %q. \n%d comparisons were done.", word,
-// 			r.resultPosition, text, r.numberOfComparison)
-// 	}
-// }
+		if word[j] == text[i+j] {
+			j++
+			if j == len(word) {
+				matches = append(matches, i)
 
-// Kmp Function kmp performing the Knuth-Morris-Pratt algorithm.
-// Prints whether the word/pattern was found and on what position in the text or not.
-// m - current match in text, i - current character in w, c - amount of comparisons.
-func Kmp(text string, word string) Result {
-	m, i, c := 0, 0, 0
-	t := kmpTable(word)
-	for m+i < len(text) {
-		fmt.Printf("\n   comparing characters %c %c at positions %d %d", text[m+i], word[i], m+i, i)
-		c++
-		if word[i] == text[m+i] {
-			fmt.Printf(" - match")
-			if i == len(word)-1 {
-				return Result{
-					m, c,
-				}
+				i = i + j
+				j = 0
 			}
-			i++
 		} else {
-			m = m + i - t[i]
-			if t[i] > -1 {
-				i = t[i]
+			i = i + j - patternTable[j]
+			if patternTable[j] > -1 {
+				j = patternTable[j]
 			} else {
-				i = 0
+				j = 0
 			}
 		}
 	}
-	return Result{notFoundPosition,
-		c,
-	}
+	return matches
 }
 
-// Table building algorithm.
-// Takes word to be analyzed and table to be filled.
-func kmpTable(word string) (t []int) {
-	t = make([]int, len(word))
-	pos, cnd := 2, 0
-	t[0], t[1] = -1, 0
-	for pos < len(word) {
-		if word[pos-1] == word[cnd] {
-			cnd++
-			t[pos] = cnd
-			pos++
-		} else if cnd > 0 {
-			cnd = t[cnd]
-		} else {
-			t[pos] = 0
-			pos++
+// Table building for kmp algorithm.
+func Table(w string) []int {
```

Comment:
I think this shouldn't be exported. Is there a reason why this is exported?

Reply:
@tjgurwara99 
Thanks for review!

You are right, I thought it did not need to be made public too, so I changed it to private!
[78ac44c](https://github.com/TheAlgorithms/Go/pull/586/commits/78ac44cafb45b89d7436fca7a65abe18b297d95c)